src/main/java/frc/robot/subsystems/drive/Drivetrain.java

```diff
@@ -1,16 +1,329 @@
+            DriveConstants.kDriveKinematics,
+            Rotation2d.fromRadians(-getGyroAngle().in(Units.Radians)),
+            new SwerveModulePosition[] {
+              m_frontLeft.getPosition(),
+              m_frontRight.getPosition(),
+              m_rearLeft.getPosition(),
+              m_rearRight.getPosition(),
+            });
+
+    m_powerDistribution.clearStickyFaults();
+    SmartDashboard.putNumber("driveVelocity", 0);
+  }
+
+  /** runs the periodic functionality of the drivetrain */
   @Override
   public void periodic() {
-    // This method will be called once per scheduler run
+    // Update the odometry in the periodic block
+    m_odometry.update(
+        Rotation2d.fromRadians(-getGyroAngle().in(Units.Radians)),
+        new SwerveModulePosition[] {
+          m_frontLeft.getPosition(),
+          m_frontRight.getPosition(),
+          m_rearLeft.getPosition(),
+          m_rearRight.getPosition(),
+        });
+
+    double ang = getGyroAngle().in(Units.Radians);
+    SmartDashboard.putNumber("delta heading", ang - m_prevAngleRadians);
+
+    m_prevAngleRadians = ang;
+
+    SmartDashboard.putNumber("heading", ang - m_headingOffsetRadians);
+
+    SmartDashboard.putNumber("right stick angle", m_rightAngGoalRadians);
+    SmartDashboard.putNumber("turn direction", m_turnDirRadians);
+  }
+
+  /**
+   * Resets the odometry to the specified pose.
+   *
+   * @param pose The pose to which to set the odometry.
+   */
+  public void resetOdometry(Pose2d pose) {
+    m_odometry.resetPosition(
+        Rotation2d.fromRadians(-getGyroAngle().in(Units.Radians)),
+        new SwerveModulePosition[] {
+          m_frontLeft.getPosition(),
+          m_frontRight.getPosition(),
+          m_rearLeft.getPosition(),
+          m_rearRight.getPosition(),
+        },
+        pose);
+  }
+
+  /**
+   * drives the drivatrain using the given inputs the magnitude of the joystick components shouldn't
+   * be > 1 (x^2 + y^2 <= 1)
+   *
+   * @param xSpeed the x-pos of the left joystick (-1, 1)
+   * @param ySpeed the y-pos of the left joystick (-1, 1)
+   * @param xRot the x-pos of the right joystick (-1, 1)
+   * @param yRot the x-pos of the right joystick (-1, 1)
+   * @param altDrive whether or not to use the alternative turning mode
+   * @param centerGyro whether or not to reset the gyro position to the current rotation
+   */
+  public void drive(Vector spdVec, Vector rotVec, boolean altDrive, boolean centerGyro) {
+    if (centerGyro) zeroHeading();
+    if (altDrive) {
+      altDrive(spdVec, rotVec);
+    } else {
+      mainDrive(spdVec, rotVec.x());
+    }
+  }
+
+  /**
+   * moves the drivetrain using the main turning mode
+   *
+   * @param xSpeed the proportion of the robot's max velocity to move in the x direction
+   * @param ySpeed the proportion of the robot's max velocity to move in the y direction
+   * @param xRot the speed to rotate with (-1, 1)
+   */
+  private void mainDrive(Vector spdVec, double xRot) {
```

Comment:
Why do we have drive, mainDrive, altDrive, move?

Reply:
drive checks if the robot is field-oriented or driver oriented, and move actually applies calculated output to the drivetrain, depending on if slew rate limiting is needed.

Reply:
If that's the case, it's a bit of an abuse of function overloading. Field and driver oriented drive should be different functions.